### PR TITLE
HIVE-23493: Rewrite plan to join back tables with many projected columns joined multiple times

### PR DESCRIPTION
**Pull request by kasakrisz**
Testing done:
```
mvn test -Dtest.output.overwrite -DskipSparkTests -Dtest=TestTezPerfConstraintsCliDriver -Dqfile=cbo_query4.q,cbo_query11.q,cbo_query74.q,query4.q,query11.q,query74.q -pl itests/qtest -Pitests
```

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * Rule to trigger {@link HiveCardinalityPreservingJoinOptimization} on top of the plan.
+ */
+public class HiveCardinalityPreservingJoinRule extends HiveFieldTrimmerRule {
+  private static final Logger LOG = LoggerFactory.getLogger(HiveCardinalityPreservingJoinRule.class);
+
+  private final double factor;
+
+  public HiveCardinalityPreservingJoinRule(double factor) {
+    super(false, "HiveCardinalityPreservingJoinRule");
+    this.factor = Math.max(factor, 0.0);
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java:40]: Maybe you should allow negative numbers (e.g., -1) to disable the optimization completely.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    JaninoRelMetadataProvider original = RelMetadataQuery.THREAD_PROVIDERS.get();
+    RelMetadataQuery.THREAD_PROVIDERS.set(getJaninoRelMetadataProvider());
+    RelMetadataQuery metadataQuery = RelMetadataQuery.instance();
+
+    RelOptCost optimizedCost = metadataQuery.getCumulativeCost(optimized);
+    RelOptCost originalCost = metadataQuery.getCumulativeCost(node);
+    originalCost = originalCost.multiplyBy(factor);
+    LOG.debug("Original plan cost {} Optimized plan cost {}", originalCost, optimizedCost);
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java:57]: nit. `Original plan cost: {} vs Optimized plan cost: {}` ? Or something like that so we can read it more clearly.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  }
+
+  @Override
+  protected RelNode trim(RelOptRuleCall call, RelNode node) {
+    RelNode optimized = new HiveCardinalityPreservingJoinOptimization().trim(call.builder(), node);
+    if (optimized == node) {
+      return node;
+    }
+
+    JaninoRelMetadataProvider original = RelMetadataQuery.THREAD_PROVIDERS.get();
```
Comment (jcamachor) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java:50]: Can we move this to the top of the method and put the `RelMetadataQuery.THREAD_PROVIDERS.set(original);` in a finally block? If there is an exception while getting the stats (for instance, if stats are not available), we should make sure that we are setting it back to the original one.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java (hunk 1 above)
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveCardinalityPreservingJoinRule.java:40]: `HiveCost` doesn't allow negative numbers:
https://github.com/apache/hive/blob/e74029d4fd5c4bfc50d33a8f1155ffacc151ba8f/ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/cost/HiveCost.java#L78

But there is a check in `CalcitePlanner`: the rule is not added if the factor is 0 or negative
```
if (factor > 0.0) {
        generatePartialProgram(program, false, HepMatchOrder.TOP_DOWN,
            new HiveCardinalityPreservingJoinRule(factor));
      }
```

**Pull request merged**